File: src/main/java/io/github/rcarlosdasilva/cogen/core/Renderer.java
package io.github.rcarlosdasilva.cogen.core;

import com.google.common.collect.Maps;
import io.github.rcarlosdasilva.kits.string.TextHelper;
import org.apache.velocity.Template;
import org.apache.velocity.VelocityContext;
import org.apache.velocity.app.Velocity;

import java.io.StringWriter;
import java.util.Map;
import java.util.Properties;

/**
 * @author <a href="mailto:rcarlosdasilva@qq.com">Dean Zhao</a>
 */
public class Renderer {

  private static Map<String, Template> templates = Maps.newHashMap();

  public static void init(String templateDir) {
    Properties properties = new Properties();
    properties.setProperty("file.resource.loader.path", templateDir);
    Velocity.init(properties);
  }

  public static String render(String path, Map<String, Object> data) {
    VelocityContext context = new VelocityContext();
    data.forEach((k, v) -> context.put(k, v));
    context.put("TextHelper", TextHelper.class);

    Template template = template(path);
    StringWriter sw = new StringWriter();
    template.merge(context, sw);
    return sw.toString();
  }

  private static Template template(String path) {
    Template template = templates.get(path);
    if (template == null) {
      template = Velocity.getTemplate(path);
      templates.put(path, template);
    }
    return template;
  }

}

File: src/main/kotlin/io/github/rcarlosdasilva/cogen/core/Renderer.kt
package io.github.rcarlosdasilva.cogen.core

import com.google.common.collect.Maps
import io.github.rcarlosdasilva.kits.string.TextHelper
import org.apache.velocity.Template
import org.apache.velocity.VelocityContext
import org.apache.velocity.app.Velocity
import java.io.StringWriter
import java.util.*

object Renderer {

  private val templates = Maps.newHashMap<String, Template>()

  fun init(templateDir: String) {
    val properties = Properties()
    properties.setProperty("file.resource.loader.path", templateDir)
    Velocity.init(properties)
  }

  fun render(path: String, data: Map<String, Any>): String {
    val context = VelocityContext()
    data.forEach { k, v -> context.put(k, v) }
    context.put("TextHelper", TextHelper::class.java)

    val template = template(path)
    val sw = StringWriter()
    template.merge(context, sw)
    return sw.toString()
  }

  private fun template(path: String): Template =
    templates[path] ?: run {
      templates[path] = Velocity.getTemplate(path)
      templates[path]!!
    }

}